File: SourceCode/Python/ChalmersGU/LinkedStack.py

from API import Stack

#/* *** ODSATag: LinkedStack *** */
#/* *** ODSATag: LinkedStackInit *** */
class LinkedStack(Stack):
    def __init__(self):
        self._top = None      # Pointer to top of stack
        self._stackSize = 0   # Size of stack
#/* *** ODSAendTag: LinkedStackInit *** */

#/* *** ODSATag: LinkedStackPush *** */
    def push(self, x):
        self._top = LinkedStackNode(x, self._top)
        self._stackSize += 1
#/* *** ODSAendTag: LinkedStackPush *** */

#/* *** ODSATag: LinkedStackPeek *** */
    def peek(self):
        if not (self._stackSize > 0): raise IndexError("peek from empty stack")
        return self._top.elem
#/* *** ODSAendTag: LinkedStackPeek *** */

#/* *** ODSATag: LinkedStackPop *** */
    def pop(self):
        if not (self._stackSize > 0): raise IndexError("pop from empty stack")
        removed = self._top
        self._top = removed.next
        removed.next = None   # For garbage collection
        self._stackSize -= 1
        return removed.elem
#/* *** ODSAendTag: LinkedStackPop *** */

    def isEmpty(self):
        return self._stackSize == 0

    def size(self):
        return self._stackSize

#/* *** ODSATag: LinkedStackIterator *** */
    def __iter__(self):
        current = self._top
        while current is not None:
            yield current.elem
            current = current.next
#/* *** ODSAendTag: LinkedStackIterator *** */


#/* *** ODSATag: LinkedStackNode *** */
# Python does not have internal classes, so we have to make the stack node standalone.
class LinkedStackNode:
    def __init__(self, elem, next):
        self.elem = elem   # Value for this node
        self.next = next   # Pointer to next node in stack
#/* *** ODSAendTag: LinkedStackNode *** */
#/* *** ODSAendTag: LinkedStack *** */


#######################################################################################
## What comes below is purely for debugging and testing purposes - it can be removed ##

def _printList(l):
    print("[", " ".join(str(e) for e in l), "]", l.size())

if __name__ == '__main__':
    a = LinkedStack()
    for i in range(23):
        a.push(chr(i+65))
        if a.size() % 5 == 0:
            _printList(a)
    _printList(a)
    while not a.isEmpty():
        assert a.peek() == a.pop(), (a,)
        if a.size() % 3 == 2:
            _printList(a)
    _printList(a)

File: SourceCode/Java_Generic/ChalmersGU/LinkedStack.java

import java.util.Iterator;

/* *** ODSATag: LinkedStack *** */
/* *** ODSATag: LinkedStackInit *** */
class LinkedStack<E> implements Stack<E> {
    private Node top;        // Pointer to top of stack
    private int stackSize;   // Size of stack

    LinkedStack() {
        top = null;
        stackSize = 0;
    }
/* *** ODSAendTag: LinkedStackInit *** */

/* *** ODSATag: LinkedStackNode *** */
    private class Node {
        E elem;       // Value for this node
        Node next;    // Pointer to next node in stack

        Node(E elem, Node next) {
            this.elem = elem;
            this.next = next;
        }
    }
/* *** ODSAendTag: LinkedStackNode *** */

/* *** ODSATag: LinkedStackPush *** */
    public void push(E x) {
        top = new Node(x, top);
        stackSize++;
    }
/* *** ODSAendTag: LinkedStackPush *** */

/* *** ODSATag: LinkedStackPeek *** */
    public E peek() {
        if (!(stackSize > 0)) throw new IndexOutOfBoundsException("peek from empty stack");
        return top.elem;
    }
/* *** ODSAendTag: LinkedStackPeek *** */

/* *** ODSATag: LinkedStackPop *** */
    public E pop() {
        if (!(stackSize > 0)) throw new IndexOutOfBoundsException("pop from empty stack");
        Node removed = top;
        top = removed.next;
        removed.next = null;   // For garbage collection
        stackSize--;
        return removed.elem;
    }
/* *** ODSAendTag: LinkedStackPop *** */

    public boolean isEmpty() {
        return stackSize == 0;
    }

    public int size() {
        return stackSize;
    }

/* *** ODSATag: LinkedStackIterator *** */
    public Iterator<E> iterator() {
        return new LinkedStackIterator();
    }

    private class LinkedStackIterator implements Iterator<E> {
        private Node current;
        LinkedStackIterator() {
            current = top;
        }
        public boolean hasNext() {
            return current != null;
        }
        public E next() {
            E x = current.elem;
            current = current.next;
            return x;
        }
    }
/* *** ODSAendTag: LinkedStackIterator *** */
/* *** ODSAendTag: LinkedStack *** */

    public void pPrint() {
        System.out.print("[ ");
        for (E e : this) System.out.print(e + " ");
        System.out.println("] " + size());
    }

    public static void main(String[] args) {
        LinkedStack<String> list = new LinkedStack<>();
        for (int i=0; i<23; i++) {
            list.push(String.valueOf((char)(i+65)));
            if (list.size() % 5 == 0) list.pPrint();
        }
        list.pPrint();
        while (!list.isEmpty()) {
            if (list.peek() != list.pop()) throw new IndexOutOfBoundsException("ERROR!!!");
            if (list.size() % 3 == 2) list.pPrint();
        }
        list.pPrint();
    }
/* *** ODSATag: LinkedStack *** */
}
/* *** ODSAendTag: LinkedStack *** */
